python/phase-1.md

```diff
@@ -0,0 +1,441 @@
+
+
+async def another_activity(arg1):
+    # Showing heartbeat and cancellation. In Python 3.7 we cannot give a
+    # specific error to Task/Future.cancel (only in 3.9). So activity
+    # cancellations appear like any other asyncio cancellations (which may be a
+    # good thing anyways).
+    try:
+        while True:
+            await asyncio.sleep(1)
+            activity.heartbeat("some details")
+    except asyncio.CancelledError as exc:
+        raise RuntimeError("someone cancelled me!") from exc
+
+
+class MyCallableClass:
+    def __call__(self):
+        pass
+
+
+class MyActivities:
+    def activity1():
+        pass
+
+
+def create_my_worker() -> temporalio.Worker:
+    return temporalio.Worker(
+        temporalio.Core.default_local(),
+        taskqueue="MyTaskQueue",
+        activities={
+            # The names can be any value
+            "simple_activity": simple_activity,
+            # Can use async function
+            "another_activity": another_activity,
+            # Can use callable class
+            "callable_class": MyCallableClass(),
+            # Can use method
+            "activity1_method": MyActivities().activity1,
+            # Can use lambda
+            "lambda": lambda arg1: arg1,
+            # Fails because not callable
+            "not_callable": MyActivities(),
+        },
+    )
+```
+
+Notes:
+
+* Cancellation is task cancellation
+  * We need to research whether, in Python 3.7, we can interrupt an async task w/ our own error reliably
+* Heartbeat is a synchronous call
+  * Should we throw a cancellation error if we know the activity is cancelled when they try to heartbeat?
+    * is this something core should be returning from `record_heartbeat` instead of lang?
+
+### Payload Converters
+
+Exported from `temporalio` (but maybe in `temporalio/converter/converter.py`, exported from `temporalio.converter`
+then exported from `temporalio`):
+
+```python
+class PayloadConverter(ABC):
+    # This can be sync or async. Always takes a list. Can return a single
+    # payload or a list of payloads with the same size as the input list or None
+    # if the value cannot be converted with this converter. Any other result
+    # causes an error. Any error raised is bubbled.
+    @abstractmethod
+    async def encode(self, *values):
+        pass
+
+    # This can be sync or async. Always takes a list. Can return a single value
+    # or a list of values with the same size as the input list or None if the
+    # value cannot be converted with this converter. Any other result causes an
+    # error. Any error raised is bubbled.
+    @abstractmethod
+    async def decode(self, *payloads):
+        pass
+
+
+class CompositePayloadConverter(PayloadConverter):
+    def __init__(self, converters) -> None:
+        pass
+
+    async def encode(self, *values):
+        pass
+
+    async def decode(self, *payloads):
+        pass
+
+
+PayloadConverter.default = CompositePayloadConverter(
```

Comment:
After thinking about this some more should Python support [pickle](https://docs.python.org/3/library/pickle.html) by default?
A lot of python devs would expect us to support this (for arbitrary object serialization).
One obvious drawback to this is that it's not a polyglot solution.

Reply:
I'm not really sure why you'd need it - unless maybe you're passing already-pickled objects from some datastore directly into activities/wfs (which I can believe).

Otherwise, if you're using our client or package, you're going to get a reasonable form of serialization for free already (with one of the defaults), so why do you care if it's pickle?

At minimum, seems like something we can add later if wanted, doesn't need to be in the proposal.

Reply:
>  One obvious drawback to this is that it's not a polyglot solution.

The most obvious. But I'll provide a pickle-based converter they can choose to use if they want.

Reply:
The problem with JSON is that by default python objects aren't serializable, only primitives are.
We can recommend using dataclasses which can be turned into dicts https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

Reply:
> The problem with JSON is that by default python objects aren't serializable, only primitives are.

In all SDKs we accept the limitation that the in/out types must be serializable by the converter by default. What happens if I have a function on a TypeScript type? Technically they could subclass `json.JSONEncoder` and impl `default` for out types, but yeah it'd just come back as the raw form of that.

>  We can recommend using dataclasses for args and return types which can be turned into dicts https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

We can do this for "encode" of the `JSONPlainPayloadConverter`, but decoded values will remain dicts. So our suggestion should really be "use dicts directly or dataclass knowing it will come in as a dict that you'll have to create your class with if needed". But I think this limitation is fine compared to, say, trying to store the class name in the payload by default (but we can definitely support that if needed, same as we support pickle but not by default).